### PR TITLE
Add config to limit maximum RMM pool size

### LINKED ISSUE #488: [FEA] Ability to limit total GPU memory used
**Is your feature request related to a problem? Please describe.**
Sometimes the plugin needs to share the GPU with another process, and the plugin's current behavior of aggressively allocating most of the GPU memory for the RMM pool is hostile to that setup.  Even if one manually configures the initial RMM pool size to be lower, it will grow without bound if the plugin requests that much memory.  It would be nice to be able to put a hard limit on the RMM memory size.

**Describe the solution you'd like**
A plugin config that can be used to set a limit on the total size of the RMM memory pool.

**Describe alternatives you've considered**
Being able to share the same RMM pool between the processes would be ideal, then we wouldn't have to hard-partition the GPU like this.  However even if that feature existed, we would still need some guarantee the plugin wouldn't end up consuming all of the shared RMM pool just as the other process needs it.

### PR DESCRIPTION
**Pull request by jlowe**
Signed-off-by: Jason Lowe <jlowe@nvidia.com>

Fixes #488.

This adds a new plugin config to artificially limit the amount of GPU memory that will be used by RMM in pool mode.
 
Note that this depends on https://github.com/rapidsai/cudf/pull/5855 and must not be merged until that change is published in the cudf-0.15-SNAPSHOT artifact.

### CODE REVIEW

**Comment by jlowe** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 1]
```diff
@@ -260,6 +260,13 @@ object RapidsConf {
     .checkValue(v => v >= 0 && v <= 1, "The fraction value must be in [0, 1].")
     .createWithDefault(0.9)
 
+  val RMM_ALLOC_MAX_FRACTION = conf("spark.rapids.memory.gpu.maxAllocFraction")
+      .doc("The fraction of total GPU memory that limits the maximum size of the RMM pool. " +
+          "If this is set to 0 then the pool size will not be artificially limited.")
+      .doubleConf
+      .checkValue(v => v >= 0 && v <= 1, "The fraction value must be in [0, 1].")
+      .createWithDefault(0)
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:268]: `createWithDefault(1)` shouldn't that naturally describe the "do not limit" case? Also if I set `RMM_ALLOC_FRACTION` to 0, it seems `RMM_ALLOC_MAX_FRACTION` should be 0 also?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 2]
```diff
@@ -260,6 +260,13 @@ object RapidsConf {
     .checkValue(v => v >= 0 && v <= 1, "The fraction value must be in [0, 1].")
     .createWithDefault(0.9)
 
+  val RMM_ALLOC_MAX_FRACTION = conf("spark.rapids.memory.gpu.maxAllocFraction")
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:263]: indentation is a little off here, or maybe it's off everywhere else ;) I believe this does follow the 4 space indentation requirement.

File: docs/configs.md [hunk 3]
```diff
@@ -31,6 +31,7 @@ Name | Description | Default Value
 -----|-------------|--------------
 <a name="memory.gpu.allocFraction"></a>spark.rapids.memory.gpu.allocFraction|The fraction of total GPU memory that should be initially allocated for pooled memory. Extra memory will be allocated as needed, but it may result in more fragmentation.|0.9
 <a name="memory.gpu.debug"></a>spark.rapids.memory.gpu.debug|Provides a log of GPU memory allocations and frees. If set to STDOUT or STDERR the logging will go there. Setting it to NONE disables logging. All other values are reserved for possible future expansion and in the mean time will disable logging.|NONE
+<a name="memory.gpu.maxAllocFraction"></a>spark.rapids.memory.gpu.maxAllocFraction|The fraction of total GPU memory that limits the maximum size of the RMM pool. If this is set to 0 then the pool size will not be artificially limited.|0.0
```
Comment (tgravescs) [inline docs/configs.md:34]: so this should be > then the memory.gpu.AllocFraction ,correct?  might be nice to reference this from the other config.

**Comment by jlowe** [comment]
build

**Comment by jlowe** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala [hunk 4]
```diff
@@ -158,6 +163,18 @@ object GpuDeviceManager extends Logging {
         logWarning(s"Initial RMM allocation(${initialAllocation / 1024 / 1024.0} MB) is " +
           s"larger than free memory(${info.free / 1024 / 1024.0} MB)")
       }
+      val maxAllocation = if (conf.rmmAllocMaxFraction < 1) {
+        (conf.rmmAllocMaxFraction * info.total).toLong
+      } else {
+        // Do not attempt to enforce any artificial pool limit based on queried GPU memory size
+        // if config indicates all GPU memory should be used.
+        0
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala:171]: We should document this

Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala:171]: Sorry I should be more clear. In the case of UVM setting it above 1.0 might be logical so we should document in the config that this is happening.

Comment (jlowe) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala:171]: I think RMM's `pool_memory_resource` will still query the GPU and use it's reported total memory as a maximum limit internally, although I'm unsure what that API does when UVM is being used.  I'll still add a comment saying the maximum limit is disabled if the specified fraction >= 1.

**Comment by jlowe** [comment]
build

**Comment by jlowe** [comment]
@revans2 this is ready for another look.

**Pull request merged**